pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java

```diff
@@ -737,17 +765,18 @@ private void initObjectTypes(Properties info) throws SQLException {
    */
   @Override
   public void close() throws SQLException {
-    if (queryExecutor == null) {
-      // This might happen in case constructor throws an exception (e.g. host being not available).
-      // When that happens the connection is still registered in the finalizer queue, so it gets finalized
-      return;
-    }
     if (queryExecutor.isClosed()) {
       return;
     }
     releaseTimer();
     queryExecutor.close();
-    openStackTrace = null;
+    LeakTracker.LeakTraceHandle<Object> leakTraceHandle = this.leakTraceHandle;
+    // Connection has been explicitly closed, unregister
+    if (leakTraceHandle != null) {
+      ConnectionLeakTrackerHolder.INSTANCE.unregister(leakTraceHandle);
+      this.leakMarker = null;
+      this.leakTraceHandle = null;
```

Comment:
Does any of this need to be in a finally block? Is there any possibility of `queryExecutor.close()` throwing an exception?

Reply:
moved above